cmd/snap/complete.go

```diff
@@ -100,11 +100,103 @@ func (s keyName) Complete(match string) []flags.Completion {
 	return res
 }
 
+type disconnectSlotOrPlugSpec struct {
+	SnapAndName
+}
+
+func (dps disconnectSlotOrPlugSpec) Complete(match string) []flags.Completion {
+	spec := &interfaceSpec{
+		SnapAndName:  dps.SnapAndName,
+		slots:        true,
+		plugs:        true,
+		connected:    true,
+		disconnected: false,
+	}
+	return spec.Complete(match)
+}
+
+type disconnectSlotSpec struct {
+	SnapAndName
+}
+
+// TODO: look at what the previous arg is, and filter accordingly
+func (dss disconnectSlotSpec) Complete(match string) []flags.Completion {
+	spec := &interfaceSpec{
+		SnapAndName:  dss.SnapAndName,
+		slots:        true,
+		plugs:        false,
+		connected:    true,
+		disconnected: false,
+	}
+	return spec.Complete(match)
+}
+
 type connectPlugSpec struct {
 	SnapAndName
 }
 
 func (cps connectPlugSpec) Complete(match string) []flags.Completion {
+	spec := &interfaceSpec{
+		SnapAndName:  cps.SnapAndName,
+		slots:        false,
+		plugs:        true,
+		connected:    false,
+		disconnected: true,
+	}
+	return spec.Complete(match)
+}
+
+type connectSlotSpec struct {
+	SnapAndName
+}
+
+// TODO: look at what the previous arg is, and filter accordingly
+func (css connectSlotSpec) Complete(match string) []flags.Completion {
+	spec := &interfaceSpec{
+		SnapAndName:  css.SnapAndName,
+		slots:        true,
+		plugs:        false,
+		connected:    false,
```

Comment:
We may want to revise this when we model one-to-many and many-to-many connections more accurately. A connected slot can still be connected to other plugs.